elixir/apps/domain/lib/domain/accounts/account/changeset.ex

```diff
@@ -2,22 +2,61 @@ defmodule Domain.Accounts.Account.Changeset do
   use Domain, :changeset
   alias Domain.Accounts.Account
 
-  def create_changeset(attrs) do
-    %Account{}
+  def changeset(account, attrs) do
+    account
     |> cast(attrs, [:name, :slug])
     |> validate_required([:name])
-    |> validate_length(:name, min: 1, max: 255)
+    |> validate_name()
+    |> trim_name_whitespace()
+    |> set_slug_default()
+    |> downcase_slug()
+    |> remove_spaces_from_slug()
     |> validate_slug()
-    |> validate_length(:slug, min: 3, max: 100)
+    |> unique_constraint(:slug, name: :accounts_slug_index)
+  end
+
+  def create_changeset(attrs) do
+    %Account{}
+    |> changeset(attrs)
+  end
+
+  defp validate_name(changeset) do
+    changeset
+    |> validate_length(:name, min: 3, max: 255)
+    |> validate_format(:name, ~r/^[a-zA-Z0-9\s_-]+$/,
+      message: "can only contain letters, numbers, spaces, underscores, and dashes"
+    )
+  end
+
+  defp trim_name_whitespace(changeset) do
+    update_change(changeset, :name, &String.trim/1)
+  end
+
+  defp set_slug_default(changeset) do
+    changeset
+    |> put_default_value(:slug, &Domain.Accounts.generate_unique_slug/0)
   end
 
   defp validate_slug(changeset) do
-    validate_change(changeset, :slug, fn field, slug ->
+    changeset
+    |> validate_length(:slug, min: 3, max: 100)
+    |> validate_format(:slug, ~r/^[a-zA-Z0-9_]+$/,
+      message: "can only contain letters, numbers, and underscores"
+    )
+    |> validate_change(:slug, fn field, slug ->
       if valid_uuid?(slug) do
-        [{field, "must can not be a valid UUID"}]
+        [{field, "can not be a valid UUID"}]
```

Comment:
😆  nice edge case catch

Reply:
Ah wait, that was there before. nit: minor grammar fix
```suggestion
        [{field, "cannot be a valid UUID"}]
```

Reply:
Nice catch!  I saw that and forgot to go back and update it

Reply:
Has been updated.